modules/ldap/src/main/java/com/neighborhoods/wasabi/ldap/impl/LdapDelegate.java

```diff
@@ -0,0 +1,857 @@
+     *
+     * @return the ldap info connection
+     * @throws LdapException the ldap exception
+     */
+    protected LdapConnection getLdapInfoConnection() throws LdapException {
+        LdapConnection connection = getLdapConnection();
+        connection.bind(this.config.getInfoDN(), this.config.getInfoPassword());
+        return connection;
+    }
+
+    /**
+     * Gets the cached user search request (and builds it if necessary).
+     *
+     * @return the user cache search request
+     * @throws LdapException If there are DN validation errors or invalid filters
+     */
+    protected SearchRequest getUserCacheSearchRequest() throws LdapException {
+        // This should never be null unless it is the first request
+        if (userCacheSearchRequest == null) {
+            // In unlikely event multiple threads reach this point, lock initialization
+            synchronized (this) {
+                SearchRequest searchRequest = new SearchRequestImpl();
+                searchRequest.setScope(SearchScope.SUBTREE);
+                searchRequest.addAttributes(this.config.getRoleAttribute());
+                searchRequest.addAttributes(this.config.getMembershipAttribute());
+                searchRequest.setTimeLimit(0);
+                searchRequest.setBase(new Dn(this.config.getWasabiDN()));
+                searchRequest.setFilter("(objectclass=*)");
+                userCacheSearchRequest = searchRequest;
+            }
+        }
+        return userCacheSearchRequest;
+    }
+
+    /**
+     * Validate a directory token
+     * 
+     * @param userDirectory - the calling user directory
+     * @param username - The username to validate
+     * @param encodedPassword - The encrypted password to validate
+     * @return Always returns whether the token is valid or not
+     */
+    @Override
+    public boolean isDirectoryTokenValid(CachedUserDirectory userDirectory, String username, String encodedPassword) {
+        try {
+            DirectoryUser userInfo = userDirectory.lookupDirectoryUser(username);
+            // It is possible for the password to be null if it was retrieved outside of the login flow
+            if (userInfo != null && userInfo.getPassword() != null) {
+                // Encoded passwords are only in the cache
+                return userInfo.getPassword().equals(encodedPassword);
+            }
+        } catch (AuthenticationException ae) {
+            return false;
+        }
+        return false;
+    }
+
+    /**
+     * Populates the user cache with all Wasabi users
+     * 
+     * @param The calling CachedUserDirectory instance
+     * @throws AuthenticationException If unable to connect to ldap or bad LDAP queries
+     */
+    @Override
+    public void populateUserCache(CachedUserDirectory userDirectory) throws AuthenticationException {
+        try (LdapConnection connection = getLdapInfoConnection()) {
+            if (connection.isConnected() && connection.isAuthenticated()) {
```

Comment:
Non-blocking suggestion: This is getting quite a few layers deep which makes it a little more difficult to follow. Can you handle the case of (!connection.isConnected() || !connection.isAuthenticated()) here so the rest of the code isn't wrapped in the if statement? Maybe also handle (response !instanceof SearchResultEntry) below as well.